File: app/src/main/java/com/nobodysapps/septimanapp/model/storage/JsonConverter.java
package com.nobodysapps.septimanapp.model.storage;

import androidx.annotation.NonNull;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.util.Map;

@Singleton
public class JsonConverter {
    private Gson gson;

    @Inject
    public JsonConverter() {
        gson = new GsonBuilder()
//                .registerTypeAdapterFactory(RuntimeAdapterFactories.gameControllerFactory())
//                .registerTypeAdapterFactory(RuntimeAdapterFactories.actionFactory())
//                .registerTypeAdapterFactory(HanabiTypeAdapterFactory.create())
//                .enableComplexMapKeySerialization()
                .create();
    }


    @NonNull
    public <T> String toJson(T obj) {
        return gson.toJson(obj);
    }

    public <T> Map toJsonMap(T obj) {
        return fromJson(toJson(obj), Map.class);
    }

    public <T> T fromJsonMap(@NonNull Map<String, Object> map, Class<T> type) {
        return gson.fromJson(toJson(map), type);
    }

    public <T> T fromJson(String json, Class<T> type) {
        return gson.fromJson(json, type);
    }
}

File: app/src/main/java/com/nobodysapps/septimanapp/model/storage/JsonConverter.kt
package com.nobodysapps.septimanapp.model.storage

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import java.lang.reflect.Type
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class JsonConverter @Inject constructor() {
    private val gson: Gson = GsonBuilder() //                .registerTypeAdapterFactory(RuntimeAdapterFactories.gameControllerFactory())
//                .registerTypeAdapterFactory(RuntimeAdapterFactories.actionFactory())
//                .registerTypeAdapterFactory(HanabiTypeAdapterFactory.create())
//                .enableComplexMapKeySerialization()
        .create()

    fun <T> toJson(obj: T): String {
        return gson.toJson(obj)
    }

//    fun <T> toJsonMap(obj: T): Map<*, *> {
//        return fromJson<Map<*, *>>(
//            toJson(obj),
//            MutableMap::class.java
//        )
//    }
//
//    fun <T> fromJsonMap(
//        map: Map<String?, Any?>,
//        type: Class<T>?
//    ): T {
//        return gson.fromJson(toJson<Map<String, Any>>(map), type)
//    }

    fun <T> fromJson(json: String?, type: Type): T = gson.fromJson(json, type)

}